### DISCUSSION TITLE
Configure two PCF8574 for 8 LED & 8 Tactile Switch to Controlling LED ON/OFF [SOLVED]

### DISCUSSION THREAD

**Discussion by Gosokan** [arendst/Tasmota#20602] [Q&A] [answered]
Hi, 
I've compiled tasmota to add PCF8574 support.  Currently, the LED can be turned on and off via Tasmota web interface. However, the tactile switch doesn't work yet. In the web interface the status will only change momentarily from 1 to 0 when the tactile switch is pressed.

Please help so that each tactile switch can function as toggle to turn the LED on or off

![Tasmota-Web-Interface_PCF8574_1](https://github.com/arendst/Tasmota/assets/13587422/84d33e79-0000-4601-bc4e-531581b02121) ![Tasmota-Web-Interface_PCF8574_2](https://github.com/arendst/Tasmota/assets/13587422/518a2d49-98eb-4637-bfb4-85476e74303e)

![tasmota-nodemcu-8-led-8-button-pcf8574](https://github.com/arendst/Tasmota/assets/13587422/efb4e562-a988-45f2-b36b-ebfd9241d4ad)

**Comment by barbudor** [comment, marked as answer]
Compile with mode 2 by adding
```
#define USE_PCF8574_MODE2
```

See doc here https://tasmota.github.io/docs/PCF8574/#mode-2

Configure leds as Relay outputs and buttons as Buttons input
And it will work automatically

**Reply by Gosokan** [reply]
I am still using Mode 1, but the tactile switch can now work to turn the LED on and off. Later, I will replace these 8 LEDs with LOW trigger type relays. Here I added a Rule by typing commands via the web console:

`Rule1 ON PCF8574-2_INP#D0=0 DO Power1 toggle ENDON ON PCF8574-2_INP#D1=0 DO Power2 toggle ENDON ON PCF8574-2_INP#D2=0 DO Power3 toggle ENDON ON PCF8574-2_INP#D3=0 DO Power4 toggle ENDON ON PCF8574-2_INP#D4=0 DO Power5 toggle ENDON ON PCF8574-2_INP#D5=0 DO Power6 toggle ENDON ON PCF8574-2_INP#D6=0 DO Power7 toggle ENDON ON PCF8574-2_INP#D7=0 DO Power8 toggle ENDON`

I don’t know if the long rule I wrote above, can be shortened using a wildcard or not. 

Next time, I might try to recompile from the source code again to be able to use Mode 2. Would using Mode 2 be better than Mode 1?

**Reply by barbudor** [reply]
No you can't use wildcards in ruled

With mode 2 you don't need any rules

**Reply by sfromis** [reply]
Mode 2 is a great enhancement over mode 1, it just acts like normal relays/switches, when you've configured the channels. I'd not recommend anyone to use mode 1, unless they have some old setup based on that.

(And while you have limited wildcards in rule triggers, they'd not work for what you try to do)

**Reply by Gosokan** [reply]
I just tried Mode 2, and writing a template like the example in the documentation, but still didn't work. After a while Tasmota will return to default Sonoff Basic. How to write the correct template?

PCF8574 with address 0x23 is used for 8 tactile switches.
PCF8574 with address 0x22 is used for 8 LEDs.
The GPIO pins on the nodemcu: SCL = D1, and SDA = D2.

**Reply by barbudor** [reply]
Details?

**Reply by sfromis** [reply]
You posted no details of what you configured, hard to see without the template you tried with. If you have a Button input registering as "pushed" for 40 seconds, this is a recovery feature for making a total reset of Tasmota.

**Reply by Gosokan** [reply]
I am trying to create a JSON template like this: 

`{“NAME”:“PCF8574”,“GPIO”:[32,33,34,35,259,258,257,256,36,37,38,39,263,262,261,260]} `

Does this mean that buttons 1 to 8 are inverted to Vcc with pull-up resistors, and LEDs 1to 8 are inverted? 
Simply put, I just want to convert the configuration that has previously been successfully run in mode 1, to mode 2.

**Reply by barbudor** [reply]
Copying the sample template that is given in the code wihtout trying to understand it and adapt to your case is prone to failure

From your photo and description above you should have all leds/relays on the 1st PCF and all buttons on the 2nd one
So your template is wrong

**Reply by barbudor** [reply]
I don't want to give you the answer, better that you find it by yourself
But look again at the doc/code that says:
```
 * GPIO  - Sequential list of pin 1 and up with configured GPIO function
 *         Function             Code      Description
 *         -------------------  --------  ----------------------------------------
 *         None                 0         Not used
 *         Button1..32     B    32..63    Button to Gnd with internal pullup
 *         Button_n1..32   Bn   64..95    Button to Gnd without internal pullup
 *         Button_i1..32   Bi   96..127   Button inverted to Vcc with internal pullup
 *         Button_in1..32  Bin  128..159  Button inverted to Vcc without internal pullup
 *         Switch1..28     S    160..187  Switch to Gnd with internal pullup
 *         Switch_n1..28   Sn   192..219  Switch to Gnd without internal pullup
 *         Relay1..32      R    224..255  Relay
 *         Relay_i1..32    Ri   256..287  Relay inverted
 *         Output_Hi       Oh   3840      Fixed output high
 *         Output_lo       Ol   3872      Fixed output low
 ```
Your first PCF (@22) has all leds so starting the GPIO section with `32, 33, 34, 35` which describes Buttons is wrong

So what should it be ?

**Reply by Gosokan** [reply]
I restarted using just 1 PCF8574 to keep things simple. LED 1 to 4 connected to P0-P3 pin on the PCF8574 module.
So the GPIO configuration should be:

Template {"NAME":"PCF8574 Test 4LED","GPIO":[224,225,226,227,0,0,0,0]},  or maybe if inverted
Template {"NAME":"PCF8574 Test 4LEDi","GPIO":[256,257,258,259,0,0,0,0]}. I can't seem to understand it well. I go back to using mode 1 again. Thank you.

**Reply by sfromis** [reply]
Since you want all the pins on one chip to be the same, you should have 8 consecutive numbers for relays, and 8 numbers for the type of input you want, like switch or button inputs, depending on how your board works.
